wolfcrypt/src/ecc.c

```diff
@@ -5841,7 +5959,76 @@ static int wc_ecc_sign_hash_hw(const byte* in, word32 inlen,
             return err;
         }
         (void)rng;
-    #endif
+    #elif defined(WOLFSSL_XILINX_CRYPT_VERSAL)
+
+#ifdef WOLFSSL_SMALL_STACK
+        K = (byte*)XMALLOC(keysize, key->heap, DYNAMIC_TYPE_PRIVATE_KEY);
+        incopy = (byte*)XMALLOC(inlen, key->heap, DYNAMIC_TYPE_HASH_TMP);
+        if (K == NULL || incopy == NULL) {
+            XFREE(incopy, key->heap, DYNAMIC_TYPE_HASH_TMP);
+            XFREE(K, key->heap, DYNAMIC_TYPE_PRIVATE_KEY);
+            return MEMORY_E;
+        }
+#else
+        if (inlen > sizeof(incopy))
+            return ECC_BAD_ARG_E;
+#endif
+
+#if defined(WOLFSSL_ECDSA_DETERMINISTIC_K) || \
+    defined(WOLFSSL_ECDSA_DETERMINISTIC_K_VARIANT)
+        err = deterministic_sign_helper(in, inlen, key);
+        if (err)
+            return err;
+        Ksize = mp_unsigned_bin_size(key->sign_k);
+        if (Ksize > keysize) {
+            err = BUFFER_E;
+            goto error_out;
+        }
+        err = mp_to_unsigned_bin(key->sign_k, K);
+        if (err)
+            goto error_out;
+        fp_reverse(K, Ksize);
+#else
+        err = wc_RNG_GenerateBlock(rng, K, keysize);
+        if (err)
+            goto error_out;
+        /* Make sure that K is max. 521 bits */
+        if (keysize == 66)
+            K[65] &= 0x1;
+#endif
+        buf_reverse(incopy, in, inlen < keysize ? inlen : keysize);
+        WOLFSSL_XIL_DCACHE_FLUSH_RANGE(XIL_CAST_U64(incopy), keysize);
+        WOLFSSL_XIL_DCACHE_FLUSH_RANGE(XIL_CAST_U64(key->privKey), keysize);
+        WOLFSSL_XIL_DCACHE_FLUSH_RANGE(XIL_CAST_U64(K), keysize);
+
+        WOLFSSL_XIL_DCACHE_INVALIDATE_RANGE(XIL_CAST_U64(out), keysize * 2);
+
+        err = XSecure_EllipticGenerateSign(&(key->xSec.cinst),
+                                           xil_curve_type[key->dp->id],
+                                           XIL_CAST_U64(incopy), keysize,
+                                           XIL_CAST_U64(key->privKey),
+                                           XIL_CAST_U64(K),
+                                           XIL_CAST_U64(out));
+        if (err) {
+            WOLFSSL_XIL_ERROR("Generate ECC signature failed", err);
+            err = WC_HW_E;
+        }
+
+        WOLFSSL_XIL_DCACHE_INVALIDATE_RANGE(XIL_CAST_U64(out), keysize * 2);
+        fp_reverse(&out[0], keysize);
```

Comment:
Please use the `mp_` functions so it works with all our math library options (not use fast math).

Reply:
Can you use `ByteReverseWords`?

Reply:
No, because they do something different.

Reply:
I would suggest moving fp_reverse to wolfmath.c and renaming to mp_reverse. If you have trouble let me know and I'll share a patch or push to your PR.

Reply:
Done